basic-algorithms-&-datastructures/index630.js

```diff
@@ -0,0 +1,500 @@
+let newsFeedArray = []
+for (property in newsFeed) {
+  newsFeedArray = newsFeed[property]
+}
+console.log('Convert the newsFeed object property values to an array with the for...of statement:', newsFeedArray); 
+
+// Return an array with the values of the newsFeed object to the newsFeedObjValues.
+const newsFeedObjValues = Object.values(newsFeed)
+console.log('Convert the newsFeed object property values to an array with Object.values: ', newsFeedObjValues)
+
+// Return an array of the property names of newsFeed in the same order of a normal loop.
+Object.keys(newsFeed)
+
+// Select and map through the articles property of the newsFeed object.
+newsFeed.articles.map(article => article)
+
+//  Return and map an array of the property names of newsFeed with the index of property names as well as the index of map set to newsFeed.
+Object.keys(newsFeed).map(index => [index, newsFeed[index]])
+
+// Create a new Date Object with the current local time zone.
+new Date()
+
+// Return a new Date Object of the second object's date_published value in newsFeedArray
+new Date(newsFeedArray[1].date_published)
+
+//  Sort through the newsFeedArray array and subtract the date_published values in the compare function to sort the array by the most recent dates published.
+newsFeedArray.sort((a,b) => { 
+    return new Date(b.date_published) - new Date(a.date_published)})
+
+// Sort By Sections Alphabetically 
+const compareSections = (a, b) => {
+const sectionA = a.section.name
+const sectionB = b.section.name
+let result = 0
+if(sectionA > sectionB) {
+  result = 1;
+} else if(sectionA < sectionB) {
+  result = -1;
+}
+return result;
+}
+// Log The Sections Alphabetically
+console.log('Sort newsFeedArray alphabetically: ', newsFeedArray.sort(compareSections))
+
+// Search Filter for the author's name: Julie Holmies in NewsFeedArray
+newsFeedArray.filter(filterName => filterName.author.name === 'Julie Holmes')
+
+// Dynamic Search Filter
+newsFeedArray.filter(filterName => filterName.author.name === 'Julie Holmes')
+const searchString = 'Search Results By Author: '
+let author = prompt('Search Articles By Authors') 
+const searchFormula = newsFeedArray.filter(filterName => filterName.author.name.toLowerCase() === author.toLowerCase())
+
+if(searchFormula.length > 0) {
+  console.log(searchString, searchFormula)
+} else {
+  console.log(searchString, 'Sorry, but the author name you requested does not match any results.')
+}
+
+// Function to Group newsFeedArray By Sections using reduce.
+function groupBySections(arrayInput) {
+return arrayInput.reduce((accumulator, currentValue) => {
+(accumulator[currentValue.section.name] = accumulator[currentValue.section.name] || []).push(currentValue)
+return accumulator
+}, {})
+};
+
+console.log('Group newsFeedArray By Sections', groupBySections(newsFeedArray))
+
+// Starred Artciles Feature
+for(char of newsFeedArray) {
+ if(char.likes > 900) {
+  char.starred = true
+ } else {
+  char.starred = false
+ }
+}
+
+/*
```

Comment:
I wasn't able to go over every step in detail, but I think this is off to a great start. I like the way you structured the lessons here, and think that each one builds off the last quite nicely. 

One thing we can do in the next draft is to remove the console.log statements. Prompting learners to write one and check the output is great, but the console can get cluttered really quickly. Also I see what you're doing with things like the hard coding of `filter` to return just the articles by 'Julie Holmes', then moving on to a more flexible solution. That's great, and when that happens don't be afraid to prompt learners to delete what they already wrote before refactoring.

When testing the final code I noticed that the `newsFeed` array seems like it has some repeats, like the article "Will Web Development Ever Rule the World?". Some articles have the `id` property while others don't, and there seems to be fewer articles than in the CodePen link. In the final version of the project we can use separate files for the `newsFeed` object, and use `export` and `import` to make it accessible from the main file.

Another thing I noticed is the `prompt` method. For this project I don't think we need a UI, and I picture it using just the code editor and console. We could probably remove the HTML page entirely and rework that code into a function that takes a string as a parameter and returns that author's articles.

Reply:
@scissorsneedfoodtoo Thanks for your consideration and these notes! Very cool to continue building on this and looking forward to further shaping it. 

Just reviewing through your notes and yes it all makes sense and these are all improvement I can take a look at. Just from quickly going over things this morning -  Noted regarding the console.logs and the removing code prompts while refactoring to keep things more clean overall. 

I remember you mentioning the possibility of importing and exporting the newsfeed data earlier on our meeting as well and I think that would also create more organization and clarity for  project interaction as well as perhaps provide a more 'realistic' experience.

Can also make that change to the final prompt and reshape the code at the end bearing in mind a more absent approach to the UI including removing the HTML page. Makes sense.

Will go through all the other specific lesson notes in detail on the second draft as well as some proof reading and looking forward to getting something improved over to you soon.

Thanks!